Fix #1415: Improve record validation docs (#1429)
kroxylicious/kroxylicious: docs/modules/record-validation/proc-record-validation.adoc

// file included in the following:
//
// assembly-record-validation-filter.adoc

[id='proc-record-validation-{context}']
= (Preview) Setting up the Record Validation filter

.Prerequisites

* An instance of Kroxylicious.
For information on deploying Kroxylicious, see the link:{github}[samples and examples^].
* A config map for Kroxylicious that includes the configuration for creating a virtual cluster.

> * Apicurio Registry (if wanting to use Schema Validation).

Reviewer: ```suggestion
* Apicurio Registry (if wanting to use Schema validation).
```

          keyRule:
            <rule definition>                                          # <2>
          valueRule:
            <rule definition>                                          # <3>
        forwardPartialRequests: false                                  # <5>
----
<1> List of topic names to which the validation rules will be applied.
<2> Validation rules that are applied to the record's key.
<3> Validation rules that are applied to the record's value.
<4> (Optional) Default rule that is applied to any topics for which there is no explict rule defined.
<5> If `false`, any record failing validation will cause the entire produce request to be rejected.
    If `true`, only partitions within the request where all records validated successfully will be forwarded to the
    broker. The response the client will receive will be a mixed response with response returned from the broker
    for the partitions whose records passed validation and error responses for the partitions that failed validation.
    If the client is using transactions, this setting is *ignored*. Any failing record validation will
    cause the entire produce request to be rejected.

> Replace the token `<rule definition>`  with the YAML either a Schema Validation Rule or a Syntactically Correct
> JSON Validation Rule depending on your requirements.

Reviewer: ```suggestion
Replace the token `<rule definition>`  in the the YAML configuration with either a Schema validation rule or a JSON Syntax validation rule depending on your requirements.
```

Author: Done.

.Example Schema Validation Rule Definition

The Schema Validation rule validates that the key or value matches the identified schema.

[source,yaml]
----
schemaValidationConfig:
    apicurioGlobalId: 1001                                       # <1>
    apicurioRegistryUrl: http://registry.local:8080              # <2>
allowNulls: true                                                 # <3>
allowEmpty: true                                                 # <4>
----
<1> Apicurio registry global ID identifying the schema that will be enforced.
<2> Apicurio Registry endpoint.

> <3> if `true`, the validator allows keys and or values to be `null`. The default is false.
> <4> if `true`, the validator allows keys and or values to be empty. The default is false.

Reviewer: ```suggestion
<3> if `true`, the validator allows keys and or values to be `null`. The default is `false`.
<4> if `true`, the validator allows keys and or values to be empty. The default is `false`.
```

> NOTE: Schema Validation mode currently has the capability to enforce only JSON schemas (https://github.com/kroxylicious/kroxylicious/issues/1431[issue])

Reviewer: ```suggestion
NOTE: Schema validation mode currently has the capability to enforce only JSON schemas (https://github.com/kroxylicious/kroxylicious/issues/1431[issue])
```

.Example JSON Syntax Validation Rule Definition

[source,yaml]
----
syntacticallyCorrectJson:
    validateObjectKeysUnique: true                               # <1>
allowNulls: true                                                 # <2>
allowEmpty: true                                                 # <3>
----

> <1> If `true`, the validator enforces that objects keys must be unique. The default is `false`.
> <2> if `true`, the validator allows keys and or values to be `null`. The default is false.
> <3> if `true`, the validator allows keys and or values to be empty. The default is false.

Reviewer: ```suggestion
<1> If `true`, the validator enforces that objects keys must be unique. The default is `false`.
<2> if `true`, the validator allows keys and or values to be `null`. The default is `false`.
<3> if `true`, the validator allows keys and or values to be empty. The default is `false`.
```

// file included in the following:
//
// assembly-record-validation-filter.adoc

[id='proc-record-validation-{context}']

> = (Preview) Setting up the Record Validation filter

Reviewer: We need an abstract for the procedure, like the Record Encryption filter.  "This procedure describes how to set up the Record Validation filter. Provide the filter configuration and rules that the filter uses to check against Kafka record keys and values."

Author: done.

          keyRule:
            <rule definition>                                          # <2>
          valueRule:
            <rule definition>                                          # <3>
        forwardPartialRequests: false                                  # <5>
----
<1> List of topic names to which the validation rules will be applied.
<2> Validation rules that are applied to the record's key.
<3> Validation rules that are applied to the record's value.
<4> (Optional) Default rule that is applied to any topics for which there is no explict rule defined.
<5> If `false`, any record failing validation will cause the entire produce request to be rejected.
    If `true`, only partitions within the request where all records validated successfully will be forwarded to the
    broker. The response the client will receive will be a mixed response with response returned from the broker
    for the partitions whose records passed validation and error responses for the partitions that failed validation.
    If the client is using transactions, this setting is *ignored*. Any failing record validation will
    cause the entire produce request to be rejected.

> Replace the token `<rule definition>`  with the YAML either a Schema Validation rule or a JSON Syntax Validation rule depending on your requirements.

Reviewer: ```suggestion
Replace the token `<rule definition>`  in the YAML configuration with either a Schema Validation rule or a JSON Syntax Validation rule depending on your requirements.
```
